[MRG] Changes to address #273

rhaxton opened the pull request:
This is a start to address the issues raised in #273 
I have added to config options to force using only gdcm or only Pillow/jpeg_ls
If neither is set, it reverts back to the "try Pillow then GDCM" previous behavior.
Right now, the config is set by environment variables just for testing, but that won't be in the final.

At the moment, I have split the "JPEG_LS" test into two tests, one that forced gdcm and one that forces Pillow.

@darcymason , and others 
Please look this over particularly with an eye to the testing strategy of forcing gdcm or Pillow.

At the moment, if a user runs py.test with Pillow, but not gdcm, installed, then the gdcm test will fail and give the false impression that something is wrong.

I want to have it say something like "Pillow works great, you can read JPEG_LS transfer syntaxes, but GDCM wasn't tested because you don't have it installed" for the user,  but I would like make sure our CI fails if it can't test GDCM...

pep8speaks commented:
Hello @rhaxton! Thanks for submitting the PR.

 - In the file [`pydicom/config.py`](https://github.com/pydicom/pydicom/blob/029b8de96a263a1320cb0f489e0ba461456dc41c/pydicom/config.py), following are the PEP8 issues :

> [Line 52:1](https://github.com/pydicom/pydicom/blob/029b8de96a263a1320cb0f489e0ba461456dc41c/pydicom/config.py#L52): [E305](https://duckduckgo.com/?q=pep8%20E305) expected 2 blank lines after class or function definition, found 1
> [Line 92:1](https://github.com/pydicom/pydicom/blob/029b8de96a263a1320cb0f489e0ba461456dc41c/pydicom/config.py#L92): [E305](https://duckduckgo.com/?q=pep8%20E305) expected 2 blank lines after class or function definition, found 1

 - In the file [`pydicom/dataset.py`](https://github.com/pydicom/pydicom/blob/029b8de96a263a1320cb0f489e0ba461456dc41c/pydicom/dataset.py), following are the PEP8 issues :

> [Line 268:13](https://github.com/pydicom/pydicom/blob/029b8de96a263a1320cb0f489e0ba461456dc41c/pydicom/dataset.py#L268): [E722](https://duckduckgo.com/?q=pep8%20E722) do not use bare except'
> [Line 475:17](https://github.com/pydicom/pydicom/blob/029b8de96a263a1320cb0f489e0ba461456dc41c/pydicom/dataset.py#L475): [E722](https://duckduckgo.com/?q=pep8%20E722) do not use bare except'
> [Line 993:25](https://github.com/pydicom/pydicom/blob/029b8de96a263a1320cb0f489e0ba461456dc41c/pydicom/dataset.py#L993): [E722](https://duckduckgo.com/?q=pep8%20E722) do not use bare except'
> [Line 996:29](https://github.com/pydicom/pydicom/blob/029b8de96a263a1320cb0f489e0ba461456dc41c/pydicom/dataset.py#L996): [E722](https://duckduckgo.com/?q=pep8%20E722) do not use bare except'
> [Line 1010:21](https://github.com/pydicom/pydicom/blob/029b8de96a263a1320cb0f489e0ba461456dc41c/pydicom/dataset.py#L1010): [E722](https://duckduckgo.com/?q=pep8%20E722) do not use bare except'
> [Line 1013:25](https://github.com/pydicom/pydicom/blob/029b8de96a263a1320cb0f489e0ba461456dc41c/pydicom/dataset.py#L1013): [E722](https://duckduckgo.com/?q=pep8%20E722) do not use bare except'
> [Line 1017:9](https://github.com/pydicom/pydicom/blob/029b8de96a263a1320cb0f489e0ba461456dc41c/pydicom/dataset.py#L1017): [E722](https://duckduckgo.com/?q=pep8%20E722) do not use bare except'
> [Line 1077:33](https://github.com/pydicom/pydicom/blob/029b8de96a263a1320cb0f489e0ba461456dc41c/pydicom/dataset.py#L1077): [E741](https://duckduckgo.com/?q=pep8%20E741) ambiguous variable name 'I'
> [Line 1166:38](https://github.com/pydicom/pydicom/blob/029b8de96a263a1320cb0f489e0ba461456dc41c/pydicom/dataset.py#L1166): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent

 - In the file [`tests/test_filereader.py`](https://github.com/pydicom/pydicom/blob/029b8de96a263a1320cb0f489e0ba461456dc41c/tests/test_filereader.py), following are the PEP8 issues :

> [Line 41:1](https://github.com/pydicom/pydicom/blob/029b8de96a263a1320cb0f489e0ba461456dc41c/tests/test_filereader.py#L41): [E722](https://duckduckgo.com/?q=pep8%20E722) do not use bare except'
> [Line 47:1](https://github.com/pydicom/pydicom/blob/029b8de96a263a1320cb0f489e0ba461456dc41c/tests/test_filereader.py#L47): [E722](https://duckduckgo.com/?q=pep8%20E722) do not use bare except'
> [Line 106:5](https://github.com/pydicom/pydicom/blob/029b8de96a263a1320cb0f489e0ba461456dc41c/tests/test_filereader.py#L106): [E722](https://duckduckgo.com/?q=pep8%20E722) do not use bare except'
> [Line 597:82](https://github.com/pydicom/pydicom/blob/029b8de96a263a1320cb0f489e0ba461456dc41c/tests/test_filereader.py#L597): [E502](https://duckduckgo.com/?q=pep8%20E502) the backslash is redundant between brackets
> [Line 605:82](https://github.com/pydicom/pydicom/blob/029b8de96a263a1320cb0f489e0ba461456dc41c/tests/test_filereader.py#L605): [E502](https://duckduckgo.com/?q=pep8%20E502) the backslash is redundant between brackets
> [Line 613:82](https://github.com/pydicom/pydicom/blob/029b8de96a263a1320cb0f489e0ba461456dc41c/tests/test_filereader.py#L613): [E502](https://duckduckgo.com/?q=pep8%20E502) the backslash is redundant between brackets
> [Line 621:82](https://github.com/pydicom/pydicom/blob/029b8de96a263a1320cb0f489e0ba461456dc41c/tests/test_filereader.py#L621): [E502](https://duckduckgo.com/?q=pep8%20E502) the backslash is redundant between brackets
> [Line 652:38](https://github.com/pydicom/pydicom/blob/029b8de96a263a1320cb0f489e0ba461456dc41c/tests/test_filereader.py#L652): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 657:38](https://github.com/pydicom/pydicom/blob/029b8de96a263a1320cb0f489e0ba461456dc41c/tests/test_filereader.py#L657): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 664:38](https://github.com/pydicom/pydicom/blob/029b8de96a263a1320cb0f489e0ba461456dc41c/tests/test_filereader.py#L664): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 669:38](https://github.com/pydicom/pydicom/blob/029b8de96a263a1320cb0f489e0ba461456dc41c/tests/test_filereader.py#L669): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment

darcymason commented:
Thanks for digging into this, @rhaxton. 

I've spoken a few times in different conversations about some tinkering I did before around configuring how different image processors were called.  Please see my new comment in #18, [here](https://github.com/pydicom/pydicom/issues/18#issuecomment-314960744) where I've pointed to and summarized the skeleton code I had started developing.

I think your question about controlling testing could be handled that way, by setting the `config.image_handlers` list to a single item, one by one for each handler being tested.

I think it also can handle your comment:

> I want to have it say something like "Pillow works great, you can read JPEG_LS transfer syntaxes, but GDCM wasn't tested because you don't have it installed" for the user, but I would like make sure our CI fails if it can't test GDCM...

...using the message concept that each handler sends back.  As for the CI, I think each test can just force the one type.  For example, if `have_gdcm`, then expect the pixel values; if not, test for the specific message.

I think this is very much in line with where you are headed, just more generalized, allowing for future handlers, and for user choice of their order.

Have a look and let me know what you think.

glemaitre commented:
Regarding the CI you should the logic to test all possibilities.
For instance, you can try to import gdcm and if not present you should test that the error is raised and that the test pass with pillow, etc.

The CI can be set by trying to install one with gdcm one with pillow and one with numpy.
What I would need to help you is a Karnaugh map of all the possibility to be tested e.g.

+---------+--------+--------+
| numpy | gdcm | pillow |
+---------+--------+--------+
| 0          | 0        | 1        |
+---------+--------+--------+
| 1          | 0        | 1        |
+---------+--------+--------+
| 0          | 1        | 0        |
+---------+--------+--------+

Regarding the implementation, would it be possible to infer the backend to use by default without setting anything (environment variable, etc.). I would think that the pattern from `imageio` is nice whith some plugins on one side, a generic function (read_image) on the other side and a request which check which plugin is appropriate to read the image. The you can always force the backend in the read_function if desired.

This design allows you to add easily a new backend and detect easily if a plugin can do or not a certain operation (by checking if a function is existing). I agree that this is requiring a lot of work and changes but I wanted to know what are your thoughts about that.

massich commented:
@glemaitre answered here before I express my thoughts in the discussion in #407 (both discussions in the same topic). Never the less, my thoughts are in line with his comment. Everything should be tested. It doesn't matter which libraries are (or aren't) present. The strategy here is, to run a test that checks that fails when the library is not present and compare the exception to make sure that we handle properly the cases where the libraries are not there.

rhaxton commented:
OK that's great.  I am going to migrate it towards the implementation @darcymason has.  I just want to point out to @glemaitre that I haven't done a Karnaugh map in nearly 30 years!

For the tests, I will just expand the tests to include the gdcm options (Pillow -> good and Not Pillow -> exception are already there).  This will also help break up dataset.py into smaller chunks.

darcymason commented:
This looks great.  Let us know when it is close to MRG status and I'll do a more thorough review.  But so far structure and layout is really good.  I'm delighted to see the pixel data code for each handler separated out.

rhaxton commented:
@glemaitre OK,  I think it is now in a state where it mostly works.  The tests all pass on my Python2.7/Pillow/jpeg_ls/numpy environment.  The test configurations I think should be:


| Numpy | Pillow | jpeg_ls | gdcm 
| ------- | -----| ---- | -----|
| 0 | 0 |  0 |  0 |

rhaxton commented:
@darcymason @glemaitre (and others).  I finally have it running and passing tests again,  but if you could look at the two files test_gdcm_pixel_data.py and test_jpeg_ls_pixel_data.py.  For test_gcdm, I have marked the tests to run on a test-by-test basis depending if gdcm is installed or not.  For test_jpeg_ls, I have marked the tests to run on the whole class.  Both methods have their merits (or demerits), which one does the team prefer (or is there a third option)? 

I have added a verbosity option to travis to show all the skipped tests.

One thing I can figure out is that testAttributeErrorInProperty now returns one of 3 different error messages, but I can't figure out why....

glemaitre commented:
I will review that tomorrow. Ping me again if you don't have news ;)

darcymason commented on tests/test_gdcm_pixel_data.py:
```diff
@@ -0,0 +1,201 @@
+emri_big_endian_name = os.path.join(test_files, "emri_small_big_endian.dcm")
+emri_jpeg_ls_lossless = os.path.join(test_files, "emri_small_jpeg_ls_lossless.dcm")
+emri_jpeg_2k_lossless = os.path.join(test_files, "emri_small_jpeg_2k_lossless.dcm")
+color_3d_jpeg_baseline = os.path.join(test_files, "color3d_jpeg_baseline.dcm")
```
Two thoughts on these test file names:  (1) if common across different test files, perhaps should be imported from a common file to avoid code duplication., but... (2) if staying here, are these all actually used?

rhaxton replied:
Yes, I tried to pull this out a few days ago, broke everything, and reverted back.  I'll have another go at it shortly.

glemaitre replied:
With @massich, it was our thoughts when speaking a "dataset" (or testdata) subpackage which can be easily import whenever needed. Refer to [here](http://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_boston.html#sklearn.datasets.load_boston) for the functionality part (not for the naming since I got the point of the reserve dataset keyword in DICOM)

massich replied:
yes, that's something I'll work on as soon as the blocker #404 is done

darcymason commented on tests/test_gdcm_pixel_data.py:
```diff
@@ -0,0 +1,201 @@
+        with self.assertRaises((NotImplementedError, )):
+            _ = self.jpeg_ls_lossless.pixel_array
+
+    @pytest.mark.skipif(not have_gdcm_handler, reason="GDCM testing is being skipped")
```
I think ```reason="GDCM is not available in this test environment"``` is a bit clearer.  Also since the string is repeated many times, maybe it should be a module-level constant, something like `no_gdcm_message` (similarly for the other message)

darcymason commented:
I left a couple comments on the code, but they may depend somewhat on your bigger question of structure for skipping.  At first glance, i prefer the skipping of the entire class, but I'm curious as to what the disadvantages of that method are.  A variation: maybe it would be even more condensed to just put all skipping of one type in the same class, rather than separate classes for different compression types?

rhaxton commented:
I prefer skipping at the class level as well.  I think that the fewer skip conditionals, the better, but right now that implies copying the whole test class with setup/teardown etc.  

Your alternative suggestion sounds like a test file for each supported configuration (numpy yes/no, pillow yes/no, gdcm yes/no, etc.)   I don't know how that would look.

If it is OK with you, I will proceed to redo the other two test files to skip by class and then make sure travis has a full configuration running to verify that we are getting closer to the test coverage we want.   After it all works again, then I will take a look at the test-by-config approach...

darcymason commented:
> If it is OK with you, I will proceed to redo the other two test files to skip by class and then make sure travis has a full configuration running to verify that we are getting closer to the test coverage we want. After it all works again, then I will take a look at the test-by-config approach...

Yes, that's perfectly fine. For me it is only a slight preference of one form over the other, just trying to minimize the number of double negatives to absorb  ;-).   But you are right, it is better to march ahead and get it working right, then worry about fine-tuning if it is even needed.

glemaitre commented on build_tools/travis/install.sh:
```diff
@@ -38,13 +38,20 @@ if [[ "$DISTRIB" == "conda" ]]; then
     fi
-    if [[ "$DEPS" == "pillow" ]]; then
+    if [[ "$JPEG2000" == "true" ]]; then
+        echo cant apt-get install libopenjp2-7 libopenjp2-7-dev
```
if you need apt-get, you need to move it in .travis.yml:

https://github.com/pydicom/pydicom/blob/master/.travis.yml#L39

glemaitre commented on build_tools/travis/install.sh:
```diff
@@ -38,13 +38,20 @@ if [[ "$DISTRIB" == "conda" ]]; then
     conda create -n testenv --yes python=$PYTHON_VERSION pip
     source activate testenv
-    conda install --yes nose pytest pytest-cov
+    conda install --yes nose pytest pytest-cov python-dateutil
```
is python-dateutil a dependencies or just optional as well?

glemaitre replied:
Checking a bit more we should change tzoffet by datetime.timedelta from the standard library

glemaitre commented on tests/test_gdcm_pixel_data.py:
```diff
@@ -0,0 +1,270 @@
+            _ = self.color_3d_jpeg.pixel_array
+
+
+@pytest.mark.skipif(have_gdcm_handler, reason=gdcm_present_message)
```
I am not sure regarding the switch. I would think about switching strategy:

```python

if has_gdcm:
    # check foncttionalities
else:
    # check error raise and not skip it

```

vsoch commented:
![image](https://user-images.githubusercontent.com/814322/28385596-93b1dcf6-6c7d-11e7-98bd-3645d4a85fa9.png)

>> I'm afraid I can't do that for you, @rhaxton ...

![image](https://user-images.githubusercontent.com/814322/28385586-87b00ce8-6c7d-11e7-962b-179ce9dd1f12.png)

rhaxton commented:
OK, I think travis and I have reached an understanding.

This PR changes the following:

- It addresses the changes in PR#401

- Added pydicom.config.image_handlers
    This is an ordered list that the dataset._get_pixel_array() method will try.
    The first one that both announces that it supports the transfer syntax and does not throw an exception, either in getting the data or reshaping it to the correct dimensions, wins.
    If they all fail, the last one to throw an exception gets to see its exception thrown up
    If no one throws an exception, but they all refuse to support the transfer syntax, then this fact is announced in a NotImplementedError exception.

- Refactored the built in image handlers into pixel_data_handlers/{handler_files}.py
   Each image handler supports two methods:
 - - handler.supports_transfer_syntax(dicom_dataset)
        Returns True if it might support this image type,  False if it will certainly not
 - - handler.get_pixeldata(dicom_dataset)
         Always tries to return a correctly sized (not necessarily correctly shaped though) numpy array of pixel data.

- pydicom.Dataset now has a _reshape_pixel_array(self, pixel_array) method that tries to reshape the given numpy array and returns the reshaped version.

- Travis now has a set of testing configurations.  I could not get libopenjp2-7-dev or CharPyLS to successfully install to allow testing, but at least when they are missing, the testing is correct.
Travis also announces when tests are skipped and shows expected failures.

- GDCM is being tested but has some unexpected failures.  On 3D datasets, it seems to only get one slice. These are marked as expected failures.

- The tests for the handlers were also moved from test_filereader.py to separate test files for each.

Things still to do:

- Get the pip install CharPyLS to work on Travis
- Get apt-get install libopenjp2-7-dev to work on Travis
- Adjust the test configurations once these are working to get test coverage

- Refactor test files out of the tests.
- Move supported transfer syntax UIDs from uid.py to the image_handler.py
- Add doc strings to the new methods.

I am at the point where I would like to merge this as it is working, and then have someone else address the travis configuration.   After that, we can actually move on to really addressing #273.

massich commented on pydicom/encaps.py:
```diff
@@ -22,7 +22,7 @@
 # or it may have a single pointer (0).
 
-from pydicom.config import logger
+import pydicom.config
```
Keep the previous call. Otherwise the calls become unreadable. 
```py
logger.debug ( "bla bla bla") 
```
is much more readable than 
```py
pydicom.config.logger.debug
```
If you have more than one logger, it might make sense. But I would still prefer something like:

```py
from moduleA.subA.sub.sub import logger as Alogger
from moduleB.sub.sub import logger as Blogger
```

rhaxton replied:
Yes, I agree.  I had to change it because we now have a circular reference with pydicom.config.  It now tries to figure out what image handlers are available by trying to import them.  Those image handlers now include pydicom.encaps since I added enhanced MR support.  
Circular references aren't the end of the world, but it does mean I have to wait until inside the method before I alias the logger now.

massich commented:
Good !! our first fully green travis in a while :)

rhaxton commented:
@massich I am no longer getting any pep8speaks reviews.  No idea why...

vsoch commented:
@rhaxton the original message he posted gets updated (see date at the bottom of this long thing) https://github.com/pydicom/pydicom/pull/409#issuecomment-314858996

rhaxton commented:
@vsoch Thanks for pointing that out.  I cleaned up the pep8 notes now.

massich commented on pydicom/encaps.py:
```diff
@@ -69,6 +69,8 @@ def defragment_data(data):
     """Read and return a single Item in the
     fragmented data stream"""
+
+    logger = pydicom.config.logger
```
I find it a bit nasty. But ok. Can you open an issue so that we can get back to this circular dependency. Maybe we can figure a nicer manner to do this.

massich commented:
@rhaxton circle CI breaks.  I think that the problem is that now, you require gdcm.

rhaxton commented:
@massich So, now I learned you have to log into circleci to see the log.txt artifact.  Looking at that I see that the import errors for jpeg_ls and gdcm are displayed because I have the lines:

`logger.debug("Could not import jpeg_ls", exc_info=e)`
and
`logger.debug("Could not import gdcm", exc_info=e)`

So, when debug logging is on (as it seems to be), I print the stack trace.

The last line of circleci does a grep for this to determine if it fails.

`grep -q "Traceback (most recent call last):"`

I removed the stack traces so cirlceci passes, but why was debug mode set in the circleci build? I could not find where this was set.  I think if debug is turned on, we should be able to show stack traces even on non-failing errors...

darcymason commented on pydicom/config.py:
```diff
@@ -71,6 +72,41 @@ def DS_decimal(use_Decimal_boolean=True):
 
 
+image_handlers = []
+
```
One last small thing... I think we should have a doc string for `image_handlers`, so it can show up in reference documentation.  My [sample code](https://github.com/pydicom/pydicom/issues/18#issuecomment-314960744) way back in the discussion had a doc string that could be a starting point, although I think it is too long.  The main point is that the user can control the order that handlers (or which handlers) are called, if they wish, by changing the list.

vsoch commented on pydicom/config.py:
```diff
@@ -73,6 +73,34 @@ def DS_decimal(use_Decimal_boolean=True):
+PixelData attribute.
+If a handler lacks required dependencies or can not otherwise be loaded,
+it shall through an ImportError.
+Handers shall have two methods:
```
throw an ImportError

vsoch commented on pydicom/config.py:
```diff
@@ -73,6 +73,34 @@ def DS_decimal(use_Decimal_boolean=True):
+
+If they all fail, the last one to throw an exception gets to see its
+exception thrown up.
+
```
hehe, should we say raised?

vsoch replied:
Although a handler throwing up an exception is quite an interesting mental visual :)

massich commented:
>I removed the stack traces so cirlceci passes, but why was debug mode set in the circleci build? I could not find where this was set. I think if debug is turned on, we should be able to show stack traces even on non-failing errors...

@rhaxton what is making circleci fail is that the gdcm, and other dependencies needed for building the documentation are not there. You should add them. Take a look at `circle.yml` you'll see that ` ./build_tools/circle/build_doc.sh` is called and the dependencies are installed there.

glemaitre commented on pydicom/tests/test_gdcm_pixel_data.py:
```diff
@@ -0,0 +1,343 @@
+save_dir = os.getcwd()
+
+
+@pytest.mark.skipif(have_gdcm_handler, reason=gdcm_present_message)
```
I am putting a comment here but this is for all the test which are skipped.
I think that the test should never be skipped.

instead we should check that something is raised. 
```python
def test_blabla():
    if have_gdcm_handler:
        # check the outpout
    else:
        # check the raised error
```

rhaxton replied:
last week @darcymason and I looked at a few approaches to this.  The old way was what you describe, but I could not figure out a way for pytest to actually announce what configuration was being tested.  The skip decorators at least allow pytest to announce "GDCM is not installed - blah blah".  If there is another way to convey that information, then these tests could be much simplified...

glemaitre replied:
I would say that we don't care so much. One of the travis build will failed and we will know the configuration of this build and the error that is raised for a failing test.
I don't see what would be the issue with that?

glemaitre commented:
@pydicom/pydcom-fff @pydicom/pydicom-core I consider this PR as a blocker since it has the key to get a green travis.

We might separate part of the PR if we want to go faster: Fix the CI and the test in a 1st PR and in a 2nd PR refactoring the pixel_data_handlers has this is done here.

WDYT?

mrbean-bremen commented on pydicom/dataset.py:
```diff
@@ -1071,22 +690,37 @@ def _get_pixel_array(self):
+                if last_exception:
+                    raise last_exception
+                else:
+                    msg = "No available image handler could " \
```
Same here. Have been bitten too often by similar backslashes (though that was in C...)

massich replied:
Same here. I'm not fan of '\' nor : 
```py
msg = "first part of the message"
msg += "second part of the message"
bla bla
msg += "more message stuff"
```

glemaitre replied:
I am not a fan of concatenating since that you repeat `msg` for useless verbosing

rhaxton replied:
Yes, I go back and forth on this one.  The slash is just not nice looking and will break if there is stuff after it, but it does break (rather than silently do something wrong).
The '(' blocking looks nicer, but it is also a tuple indicator and it is hard to tell:

```
msg = ('try this',
       'and this')
```
from
```
msg = ('try this'
       'and this')
```
worse, it is hard to tell which is intended.  With the slash outside the quotes, I believe the intent is always clear (is there any instance where the slash outside the quotes might mean something else even with stuff after the slash?).

all that being said, I'll do what everyone else wants to do for this...

rhaxton commented:
@glemaitre @darcymason 
In my comment above last week I moved that we merge this as is and take my "to-do" list into different issues (like getting the dependencies working in travis, fixing the doc strings, etc.) 

@massich reviewed, provided some good advice, and approved. I think we should merge.

Also, when I changed the title to [MRG], the email chain still ended up under the original [WIP] in gmail.  No idea why.

massich commented:
@rhaxton I aproved those changes, but not the PR. There are `+2,279 −787 ` diff lines in `18` files. This will take several iterations and lots of time to review. Which is fine. 

Once said that I'm fine in splitting this PR for quick merge. But this needs to be done carefully (meaning, in agreement)

glemaitre commented:
> Also, when I changed the title to [MRG], the email chain still ended up under the original [WIP] in gmail. No idea why.

Gmail keep the title when it send you the first notification. It would be annoying if it was updating since that you will get a new thread of mail each time.

rhaxton commented on pydicom/config.py:
```diff
@@ -71,6 +72,69 @@ def DS_decimal(use_Decimal_boolean=True):
 
 
+image_handlers = []
+"""Image handlers for converting pixel data.
```
@glemaitre Is there an easy way for me to see the docs generated by circleci from this build? It is not obvious from the control panel...

glemaitre replied:
you need to connect, then go in artifact /home/ubuntu/doc/_build/html/index.html

Something like that. The problem is that the docstring here will not be render anywhere.

rhaxton commented on build_tools/travis/install.sh:
```diff
@@ -38,13 +38,20 @@ if [[ "$DISTRIB" == "conda" ]]; then
+        echo cant apt-get install libopenjp2-7 libopenjp2-7-dev
+    fi
+    if [[ "$JPEG_LS" == "true" ]]; then
+        echo cant do python -m pip install CharPyLS
```
@glemaitre I don't know travis well enough to figure out what the issue is.  I deferred this to better minds.

mrbean-bremen commented:
I think there is now agreement that this shall be merged more or less as is, and the remaining issues be handled in separate PRs. The current state of the PR is certainly an improvement compared to master, so I see problem merging it.
I also really would like to see Travis fixed, and I think this PR gets just too big.

massich commented:
I gave another read to all code changes and they are kind of deep. So I don't think we can do them here. 

What about this:
@rhaxton can you make a list of the things that remain to be done (things pointed by @glemaitre or otherwise) so that we can merge this as a CI fixture?

If so, you can count me as MRG+1

cc: @pydicom/pydcom-fff

darcymason commented on pydicom/pixel_data_handlers/gdcm_handler.py:
```diff
@@ -0,0 +1,110 @@
+    have_numpy = False
+    raise
+
+have_gdcm = True
```
I'd rather use all capitals very sparingly, if at all.  For me, python's 'duck typing' concept means you don't have to embed information about a variable/constant into its name, including to me whether it is global or not.  Pep-8 does suggest constants in all capitals, but I personally don't need the 'shouting' feel when reading them.

scaramallion replied:
I think caps for globals is useful, since it tells me not to mess with their values and lets me know to look beyond local scope when trying to follow the code. Within a module it also implies that the variable is public (contrast with the `can_use_gdcm` variable on line 15, although that's probably better off being `_can_use_gdcm` or being defined within `get_pixeldata`).

darcymason commented:
I'm just confirming: I think all are agreed to merge this entirely, correct?  If I don't hear any comments, I will merge later today.

Personally I feel it is time to get this major update in place before too many collisions with other changes people are trying to make.  Actually, speaking of which, are there others PRs that should be merged first to reduce the number of conflicts?  My guess is that it would be easier to do this one first, and adapt the other (smaller) PRs.

darcymason commented:
Adding another comment to ping the FFF group, to ensure people see it:  @pydicom/pydcom-fff, please vote on my previous comment or add a response.

scaramallion commented:
I'm OK to merge this first. Its easier to change and review smaller PRs than the reverse.

darcymason commented:
Okay, merging this in...  but as @massich asked:

> @rhaxton can you make a list of the things that remain to be done (things pointed by @glemaitre or otherwise) so that we can merge this as a CI fixture?

I think that is still an outstanding task, doesn't necessarily need to be @rhaxton if someone else can summarize.
